airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/DetectStreamToFlush.java

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.integrations.destination_async;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.airbyte.integrations.destination_async.buffers.BufferDequeue;
+import io.airbyte.protocol.models.v0.StreamDescriptor;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+
+/**
+ * This class finds the best, next stream to flush.
+ */
+@Slf4j
+public class DetectStreamToFlush {
+
+  private static final double EAGER_FLUSH_THRESHOLD = 0.90;
+  private static final long QUEUE_FLUSH_THRESHOLD_BYTES = 10 * 1024 * 1024; // 10MB
+  private static final long MAX_TIME_BETWEEN_REC_MIN = 5L;
+  private final BufferDequeue bufferDequeue;
+  private final RunningFlushWorkers runningFlushWorkers;
+  private final AtomicBoolean isClosing;
+  private final DestinationFlushFunction flusher;
+
+  public DetectStreamToFlush(final BufferDequeue bufferDequeue,
+                             final RunningFlushWorkers runningFlushWorkers,
+                             final AtomicBoolean isClosing,
+                             final DestinationFlushFunction flusher) {
+    this.bufferDequeue = bufferDequeue;
+    this.runningFlushWorkers = runningFlushWorkers;
+    this.isClosing = isClosing;
+    this.flusher = flusher;
+  }
+
+  /**
+   * Get the best, next stream that is ready to be flushed.
+   *
+   * @return best, next stream to flush. If no stream is ready to be flushed, return empty.
+   */
+  public Optional<StreamDescriptor> getNextStreamToFlush() {
```

Comment:
looks like this isn't wired up in `FlushWorkers` - is the plan to do that in a follow up PR?